test/modules/TestResultCoordinator/report/DirectMethodReport/NetworkStatusTimeline.cs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TestResultCoordinator.Report.DirectMethodReport
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.NetworkController;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil.TestResults;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    class NetworkStatusTimeline
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger(nameof(NetworkStatusTimeline));
+
+        readonly IReadOnlyList<NetworkControllerTestResult> networkControllerTestResults;
+        readonly TimeSpan tolerancePeriod;
+        bool testResultsValidated = false;
+
+        public static async Task<NetworkStatusTimeline> Create(ITestResultCollection<TestOperationResult> networkControllerTestOperationResults, TimeSpan tolerancePeriod)
+        {
+            List<NetworkControllerTestResult> networkControllerTestResults = new List<NetworkControllerTestResult>();
+
+            while (await networkControllerTestOperationResults.MoveNextAsync())
+            {
+                Option<NetworkControllerTestResult> networkControllerTestResult = GetNetworkControllerTestOperationResult(networkControllerTestOperationResults.Current);
+
+                networkControllerTestResult.ForEach(
+                    r =>
+                    {
+                        networkControllerTestResults.Add(r);
+                    });
+            }
+
+            return new NetworkStatusTimeline(networkControllerTestResults, tolerancePeriod);
+        }
+
+        NetworkStatusTimeline(List<NetworkControllerTestResult> networkControllerTestResults, TimeSpan tolerancePeriod)
+        {
+            networkControllerTestResults.Sort(new NetworkTimelineResultComparer());
+            this.networkControllerTestResults = networkControllerTestResults;
+            this.tolerancePeriod = tolerancePeriod;
+        }
+
+        private void ValidateNetworkControllerTestResults()
+        {
+            for (int i = 0; i < this.networkControllerTestResults.Count; i++)
+            {
+                NetworkControllerTestResult curr = this.networkControllerTestResults[i];
+                if (NetworkControllerOperation.SettingRule.Equals(curr.Operation))
```

Comment:
invert this if statement and throw exception immediately.  I think this is more readable.
e.g.
if (condition1)
    throw ...
if (condition2)
    throw ...
...